pkg/prometheus/watcher.go

```diff
@@ -470,3 +605,105 @@ func (w *Watcher) syncAlertManager(key string) error {
 	_, err = w.mClient.MonitoringV1().Alertmanagers(am.Namespace).Update(am)
 	return err
 }
+
+func (w *Watcher) handleAlertmanagerConfigAdd(obj interface{}) {
+	key, ok := keyFunc(obj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config added: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) handleAlertmanagerConfigUpdate(oldObj, newObj interface{}) {
+	key, ok := keyFunc(newObj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config added: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) handleAlertmanagerConfigDelete(obj interface{}) {
+	key, ok := keyFunc(obj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config deleted: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) syncAlertManagerConfig(key string) error {
+	var alertManagerName string
+	var data []byte
+
+	log.Debugf("syncing alert manager config: key: %s", key)
+	obj, exists, err := w.amcInf.GetIndexer().GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		log.Error("Key not found...")
+		return nil
+	}
+
+	amc := obj.(*monitoringv1.AlertmanagerConfig)
+	if nil == amc {
+		log.Errorf("Got an invalid amc object for: %s", key)
+		return nil
+	}
+
+	log.Debugf("syncing alert manager config: key: %s, name: %s", key, amc.Spec.Type)
+
+	for key, val := range amc.Labels {
+		log.Debugf("Labels: %s %s", key, val)
+		if key == "alertmanager" {
+			alertManagerName = val
+			break
+		}
+	}
+
+	if alertManagerName == "" {
+		log.Errorf("Alert manager label missing in alertmanager config: %s", key)
+		return nil
+	}
+
+	switch amc.Spec.Type {
+	case "slack":
+		log.Debug("Formatting slack configuration")
+		data, err = w.formatAlertConfig(amc.Spec.Params)
+		if err != nil {
+			log.Errorf("Failed to format alert config for: %s", key)
+			return nil
+		}
+	default:
+		log.Errorf("Got an invalid type: %s for %s", amc.Spec.Type, key)
+		return nil
+	}
+
+	secretName := "alertmanager-" + alertManagerName
+
+	exists, _ = w.checkSecretExists(amc.Namespace, secretName)
+	if exists {
+		log.Infof("Secret for alertmanager: %s exists deleting it", key)
+		_, err = w.deleteSecret(amc.Namespace, secretName)
```

Comment:
There is no way to update/patch the secret, we need to delete and recreate it. It's best to make sure that the current state of AlertmanagerConfig is reflected on change.